woocommerce/woocommerce-ios
# Add Empty State View for Products and Orders Search

Closes #1974

⚠️ The #2070 needs to be reviewed and merged first before this. 

## Design 

Here is a video showing how it works: [21s video](https://drive.google.com/file/d/1-HBg-bD4DPGgvodbcfmx5I47kzd5Tj0a/view?usp=sharing).

Before | After (Light) | After (Dark)
--------|-------|---
<img width="545" alt="before-portrait" src="https://user-images.githubusercontent.com/198826/78278517-633bb400-74d3-11ea-912e-e2bf5feece4c.png"> |<img width="545" alt="white-portrait" src="https://user-images.githubusercontent.com/198826/78278528-6898fe80-74d3-11ea-8507-e51f3e0a5c61.png">|<img width="545" alt="dark-portrait" src="https://user-images.githubusercontent.com/198826/78278537-6c2c8580-74d3-11ea-89a2-b528c21b702d.png">

Following WCAndroid's behavior, the image is hidden when in landscape mode for iPhone. 

Before | After (Light) | After (Dark)
--------|-------|--
<img width="964" alt="before-landscape" src="https://user-images.githubusercontent.com/198826/78278691-a433c880-74d3-11ea-82aa-736106c06192.png">|<img width="964" alt="white-landscape" src="https://user-images.githubusercontent.com/198826/78278699-a85fe600-74d3-11ea-854d-feaa4db6d9ba.png">|<img width="964" alt="dark-landscape" src="https://user-images.githubusercontent.com/198826/78278706-ab5ad680-74d3-11ea-97fd-7acb48d9d6a8.png">

For large font sizes, the view will be **scrollable**.

Large Font (Portrait / Scrollable) | Large Font (Landscape / Scrollable)
--------|-------
<img width="545" alt="large-portrait" src="https://user-images.githubusercontent.com/198826/78278899-f4128f80-74d3-11ea-851f-8fb21ee96206.png">|<img width="964" alt="large-landscape" src="https://user-images.githubusercontent.com/198826/78278947-05f43280-74d4-11ea-88bf-0177470ad9b8.png">

The vertical alignment is slightly off on iPad. We'll have to do some more investigation on that. On 3.8, there's a [very noticeable gap underneath the table](https://user-images.githubusercontent.com/198826/78279161-59ff1700-74d4-11ea-83dd-0a97c7d89a18.png). This may be the cause of the bug. 


iPad (Portrait) | iPad (Landscape)
--------|-------
<img width="1046" alt="ipad-portrait" src="https://user-images.githubusercontent.com/198826/78279555-ef021000-74d4-11ea-8177-44cacaac5eb2.png">        |       <img width="1404" alt="ipad-landscape" src="https://user-images.githubusercontent.com/198826/78279540-eb6e8900-74d4-11ea-960c-09dbdc353519.png">

## Changes

The PRs #2062, #2066, and #2070 were the prerequisites for this to work. 

This works like the starter view implementation in #1804. The original `SearchUICommand.emptyStateText` requirement was removed (62dc52cefc9ff046b88f0461370d61935ab9b082). The `SearchUICommand` subclasses now need to provide a `ViewController` (`EmptyStateViewControllerType`) to use for the empty state: 

https://github.com/woocommerce/woocommerce-ios/blob/4de0d21787b57e389c8e1065d1f236e41907977d/WooCommerce/Classes/ViewRelated/Search/SearchUICommand.swift#L29-L39

The `EmptyStateViewControllerType` instance is created on demand. And is displayed as a child view controller of `SearchViewController`. 

![image](https://user-images.githubusercontent.com/198826/78281171-7ea8be00-74d7-11ea-8af3-5e85e454edd6.png)

### Updating the Message

Subclasses have the opportunity to update the `EmptyStateViewControllerType` by overriding `configureEmptyStateViewControllerBeforeDisplay`:

https://github.com/woocommerce/woocommerce-ios/blob/4de0d21787b57e389c8e1065d1f236e41907977d/WooCommerce/Classes/ViewRelated/Search/SearchUICommand.swift#L41-L52

This is where `SearchUICommand` subclasses can update the displayed message to something like "Could not find results for {search_text}". 

### Default `EmptyStateViewControllerType`

The new [`EmptySearchResultsViewController`](https://github.com/woocommerce/woocommerce-ios/blob/issue/1974-empty-state-framework/WooCommerce/Classes/ViewRelated/ReusableViews/EmptySearchResultsViewController/EmptySearchResultsViewController.swift) is the default view that will be used if the subclasses do not change the type of `EmptyStateViewControllerType`. In the end, subclasses like `ProductSearchUICommand` only need to implement message updates:

https://github.com/woocommerce/woocommerce-ios/blob/4de0d21787b57e389c8e1065d1f236e41907977d/WooCommerce/Classes/ViewRelated/Search/Product/ProductSearchUICommand.swift#L33-L43

## Testing

1. Navigate to Orders → Search
2. Enter a search string that should not return any result. 
3. Confirm that the new empty view is shown. 
4. Rotate to landscape mode.
5. Confirm that the image is hidden (if you're using an iPhone) and you can still see the message.
6. Clear the search text field. 
7. Confirm that the default view is shown. For Orders, it's the list of Statuses. For Products, it's the list of products. 

Please repeat the same testing for Products. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [ ] Merge #2070 first
- [ ] Change the target branch of this PR to `develop`
